NeoML/include/NeoML/Dnn/Dnn.h

```diff
@@ -700,6 +684,54 @@ inline CArchive& operator>>( CArchive& archive, CDnn& dnn)
 
 void NEOML_API SerializeLayer( CArchive& archive, IMathEngine& mathEngine, CPtr<CBaseLayer>& layer );
 
+//------------------------------------------------------------------------------------------------------------
+
+// This class can initialize a reference dnn, that has the same configuration as the original dnn
+// and shares parameter blobs with the original dnn to save memory.
+// Useful for multi-threaded inference where each thread can operate own reference dnn independently.
+// Learning is disabled for both the original dnn and the reference dnn.
+// Creates a copy of the original dnn's random generator to use it for inference.
+class NEOML_API CDnnReferenceRegister final {
```

Comment:
And it is becoming factory in this PR, not register.

Reply:
`factory` renamed

Reply:
Yes, it is good point, I will think how to avoid 2 copies of a dnn

Reply:
I have thought about it. As I see

**The main working scenario** is
```
{ // Somewhere somewhen
       Dnn dnn( random, mathEngine ); // A dnn was created
       for ( int i = 0; i < n; ++i ) { // and it will be trained then
             dnn.RunAndLearnOnce(); // maybe using the CDistributedTraining
       }
       // ...
       { // Myabe it will be prepared for inference
	         dnn.DeleteLayer( "labels" );
	         dnn.DeleteLayer( "loss" );

                 ( void ) OptimizeDnn( dnn );
       }
       // Then it will be dumped to a file surely to store the results of its training
       CArchive archive( &file, CArchive::store );
       dnn.Serialize( archive );
} // and dnn will be destroyed
```
With big possibility in another execution file, the dnn will be loaded again
```
{
       CArchive archive( &file, CArchive::store );
       CReferenceDnnFactory factory( cpuMathEngine, archive, /*seed*/42, /*optimizeDnn*/true );
       // No double dnns in memory
       // If the inference preparing is skipped, I have add the call optimizeDnn inside the factory ctor
       
       // Then it will be used for the multi-threaded inference only
       // ...
}
```

For the **another scenario** I have tried to add the next possibility
```
CReferenceDnnFactory factory;
{
       CDnn dnn; // is the trained dnn and is created with CPU MathEngine
       factory = new CReferenceDnnFactory( dnn ); // thief the dnn's state
       // Moving implemented using the reference dnn mechanism
} // The dnn cannot be used and should be destroyed
```

Reply:
fixed